src/main/java/ml/compute/InputData.java

```diff
@@ -0,0 +1,28 @@
+package ml.compute;
+
+import java.util.Set;
+
+import ml.preprocessor.types.PreprocessedData;
+import ml.preprocessor.types.RawData;
+
+/**
+ * Created by mikeecb on 17/10/2015.
+ */
+public class InputData {
```

Comment:
I don't think that will work. Becuase PreprocessedData and RawData should not be type paramaterized imo. so they would have to implements InputData with a certain type and that is just too restrictive. What vistor pattern gives us is the abstract of input data without saying anything about the data. Then the ML technique can decide which data it wants and throw an exception or something if the given class is not approrpiate.